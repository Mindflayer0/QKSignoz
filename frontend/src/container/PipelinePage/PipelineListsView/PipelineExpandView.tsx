import { PlusCircleOutlined } from '@ant-design/icons';
import { useIsDarkMode } from 'hooks/useDarkMode';
import React, { useCallback, useMemo } from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { useTranslation } from 'react-i18next';

import { tableComponents } from '../config';
import { ActionMode, ActionType } from '../Layouts';
import { ModalFooterTitle } from '../styles';
import {
	AlertMessage,
	PipelineColumn,
	PipelineOperators,
	ProcessorColumn,
} from '.';
import { processorColumns } from './config';
import { FooterButton, StyledTable } from './styles';
import DragAction from './TableComponents/DragAction';
import PipelineActions from './TableComponents/PipelineActions';
import { getTableColumn, getUpdatedRow } from './utils';

function PipelineExpandView({
	handleAlert,
	setActionType,
	handleProcessorEditAction,
	isActionMode,
	setIsVisibleSaveButton,
	selectedPipelineDataState,
	setSelectedPipelineDataState,
	processorData,
}: PipelineExpandViewProps): JSX.Element {
	const { t } = useTranslation(['pipeline']);
	const isDarkMode = useIsDarkMode();

	const processorDeleteHandler = useCallback(
		(record: ProcessorColumn) => (): void => {
			setIsVisibleSaveButton(ActionMode.Editing);
			const filteredProcessorData = selectedPipelineDataState.operators.filter(
				(item: PipelineOperators) => item.id !== record.id,
			);
			const modifiedProcessorData = { ...selectedPipelineDataState };
			modifiedProcessorData.operators = filteredProcessorData;
			setSelectedPipelineDataState(modifiedProcessorData);
		},
		[
			selectedPipelineDataState,
			setIsVisibleSaveButton,
			setSelectedPipelineDataState,
		],
	);

	const handleProcessorDeleteAction = useCallback(
		(record: ProcessorColumn) => (): void => {
			handleAlert({
				title: `${t('delete_processor')} : ${record.name}?`,
				descrition: t('delete_processor_description'),
				buttontext: t('delete'),
				onOk: processorDeleteHandler(record),
			});
		},
		[handleAlert, processorDeleteHandler, t],
	);

	const columns = useMemo(() => {
		const fieldColumns = getTableColumn(processorColumns);
		if (isActionMode === ActionMode.Editing) {
			fieldColumns.push(
				{
					title: '',
					dataIndex: 'action',
					key: 'action',
					render: (_value, record): JSX.Element => (
						<PipelineActions
							isPipelineAction={false}
							editAction={handleProcessorEditAction(record)}
							deleteAction={handleProcessorDeleteAction(record)}
						/>
					),
				},
				{
					title: '',
					dataIndex: 'dragAction',
					key: 'dragAction',
					render: () => <DragAction />,
				},
			);
		}
		return fieldColumns;
	}, [handleProcessorDeleteAction, handleProcessorEditAction, isActionMode]);

	const reorderProcessorRowData = useCallback(
		(updatedRow: PipelineOperators[]) => (): void => {
			setIsVisibleSaveButton(ActionMode.Editing);
			const modifiedProcessorData = { ...selectedPipelineDataState };
			modifiedProcessorData.operators = updatedRow;
			setSelectedPipelineDataState(modifiedProcessorData);
		},
		[
			selectedPipelineDataState,
			setIsVisibleSaveButton,
			setSelectedPipelineDataState,
		],
	);

	const onCancelProcessorReorder = useCallback(
		() => (): void => {
			setSelectedPipelineDataState(selectedPipelineDataState);
		},
		[selectedPipelineDataState, setSelectedPipelineDataState],
	);

	const moveProcessorRow = useCallback(
		(dragIndex: number, hoverIndex: number) => {
			if (
				selectedPipelineDataState.operators &&
				isActionMode === ActionMode.Editing
			) {
				const updatedRow = getUpdatedRow(
					selectedPipelineDataState.operators,
					dragIndex,
					hoverIndex,
				);

				handleAlert({
					title: t('reorder_processor'),
					descrition: t('reorder_processor_description'),
					buttontext: t('reorder'),
					onOk: reorderProcessorRowData(updatedRow as PipelineOperators[]),
					onCancel: onCancelProcessorReorder(),
				});
			}
		},
		[
			t,
			handleAlert,
			isActionMode,
			onCancelProcessorReorder,
			selectedPipelineDataState.operators,
			reorderProcessorRowData,
		],
	);

	const onClickHandler = useCallback((): void => {
		setActionType(ActionType.AddProcessor);
	}, [setActionType]);

	const footer = useCallback((): JSX.Element | undefined => {
		if (isActionMode === ActionMode.Editing) {
			return (
				<FooterButton type="link" onClick={onClickHandler}>
					<PlusCircleOutlined />
					<ModalFooterTitle>{t('add_new_processor')}</ModalFooterTitle>
				</FooterButton>
			);
		}
		return undefined;
	}, [onClickHandler, t, isActionMode]);

	return (
		<DndProvider backend={HTML5Backend}>
			<StyledTable
				isDarkMode={isDarkMode}
				showHeader={false}
				columns={columns}
				size="small"
				components={tableComponents}
				dataSource={processorData}
				pagination={false}
				onRow={(
					_record: ProcessorColumn,
					index?: number,
				): React.HTMLAttributes<unknown> => {
					const attr = {
						index,
						moveRow: moveProcessorRow,
					};
					return attr as React.HTMLAttributes<unknown>;
				}}
				footer={footer}
			/>
		</DndProvider>
	);
}

interface PipelineExpandViewProps {
	handleAlert: (props: AlertMessage) => void;
	setActionType: (actionType?: ActionType) => void;
	handleProcessorEditAction: (record: ProcessorColumn) => () => void;
	isActionMode: string;
	setIsVisibleSaveButton: (actionMode: ActionMode) => void;
	selectedPipelineDataState: PipelineColumn;
	setSelectedPipelineDataState: (data: PipelineColumn) => void;
	processorData: Array<ProcessorColumn> | undefined;
}

export default PipelineExpandView;
