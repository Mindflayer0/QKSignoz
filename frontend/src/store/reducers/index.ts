import { combineReducers } from 'redux';

import appReducer from './app';
import dashboardReducer from './dashboard';
import globalTimeReducer from './global';
import { LogsReducer } from './logs';
import metricsReducers from './metric';
import { PipelineReducer } from './pipeline';
import { ServiceMapReducer } from './serviceMap';
import traceReducer from './trace';
import { usageDataReducer } from './usage';

const reducers = combineReducers({
	traces: traceReducer,
	usageDate: usageDataReducer,
	globalTime: globalTimeReducer,
	serviceMap: ServiceMapReducer,
	dashboards: dashboardReducer,
	app: appReducer,
	metrics: metricsReducers,
	logs: LogsReducer,
	pipeline: PipelineReducer,
});

export type AppState = ReturnType<typeof reducers>;

export default reducers;
