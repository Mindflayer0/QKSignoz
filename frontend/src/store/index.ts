import { applyMiddleware, compose, createStore } from 'redux';
import type { ThunkMiddleware } from 'redux-thunk';
import thunk from 'redux-thunk';

import reducers, { AppState } from './reducers';

const composeEnhancers =
	(window && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const store = createStore(
	reducers,
	// @TODO Add Type for AppActions also
	composeEnhancers(applyMiddleware(thunk as ThunkMiddleware<AppState>)),
);

export default store;
