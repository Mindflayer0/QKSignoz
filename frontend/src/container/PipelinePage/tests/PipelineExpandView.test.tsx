import { render } from '@testing-library/react';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import i18n from 'ReactI18';
import store from 'store';

import PipelineExpandView from '../PipelineListsView/PipelineExpandView';

beforeAll(() => {
	Object.defineProperty(window, 'matchMedia', {
		writable: true,
		value: jest.fn().mockImplementation((query) => ({
			matches: false,
			media: query,
			onchange: null,
			addListener: jest.fn(),
			removeListener: jest.fn(),
			addEventListener: jest.fn(),
			removeEventListener: jest.fn(),
			dispatchEvent: jest.fn(),
		})),
	});
});

describe('PipelinePage', () => {
	it('should render PipelineExpandView', () => {
		const handleAlert = jest.fn();
		const setProcessorDataSource = jest.fn();
		const setActionType = jest.fn();
		const handleProcessorEditAction = jest.fn();
		const { asFragment } = render(
			<MemoryRouter>
				<Provider store={store}>
					<I18nextProvider i18n={i18n}>
						<PipelineExpandView
							handleAlert={handleAlert}
							processorDataSource={[]}
							setProcessorDataSource={setProcessorDataSource}
							setActionType={setActionType}
							handleProcessorEditAction={handleProcessorEditAction}
						/>
					</I18nextProvider>
				</Provider>
			</MemoryRouter>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
