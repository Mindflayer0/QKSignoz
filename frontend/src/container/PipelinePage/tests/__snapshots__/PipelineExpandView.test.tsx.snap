// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelinePage should render PipelineExpandView section 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  margin-left: 5.875rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.c0 .ant-table-tbody > tr > td {
  border: none;
}

.c0 .ant-table-tbody > tr:last-child > td {
  border: none;
}

.c0 .ant-table-content {
  background: #1d1d1d;
}

.c2 {
  font-style: normal;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

<div
    class="ant-table-wrapper c0 css-dev-only-do-not-override-1i536d8"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-1i536d8"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-small ant-table-empty"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-placeholder"
                    draggable="true"
                  >
                    <td
                      class="ant-table-cell"
                      colspan="4"
                    >
                      <div
                        class="css-dev-only-do-not-override-1i536d8 ant-empty ant-empty-normal"
                      >
                        <div
                          class="ant-empty-image"
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fill-rule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fill-rule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          class="ant-empty-description"
                        >
                          No data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            class="ant-table-footer"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-1i536d8 ant-btn-link c1"
              type="button"
            >
              <span
                aria-label="plus-circle"
                class="anticon anticon-plus-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
                  />
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                </svg>
              </span>
              <span
                class="c2"
              >
                add_new_processor
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
